### PR TITLE
[#23] [Integrate] As a user, I can see the survey details 

### PR DESCRIPTION
**Pull request by Tuubz**
https://github.com/nimblehq/ic-flutter-taher-toby/issues/23

## What happened 👀

- Integrate the survey details on the form screen:
   - Add `decodeJsonFromData` to properly decode the JSON response for survey details
   - Add VM and state handling for the form screen
- Show progress indicator while the survey details are loading
   - Change the theme color `primary`, to reflect the color on the progress indicator
- Add unit tests

## Insight 📝

Apparently `FormState` already exists, so this had to be added import `package:flutter/material.dart' hide FormState;`

## Proof Of Work 📹

**SUCCESS**

https://user-images.githubusercontent.com/18277915/232450803-a52d01df-e40c-4be9-a042-c0eff4a68e91.mov

**ERROR**

https://user-images.githubusercontent.com/18277915/232451032-8cf88ab3-e105-41ba-b412-dcd3b0a7b1b5.mov

### CODE REVIEW

File: lib/ui/form/form_screen.dart [hunk 1]
```diff
@@ -1,44 +1,126 @@
...
-    _getSurveyDetailsUseCase.call(
-      GetSurveyDetailsInput(
-        surveyId: surveyId,
-      ),
-    );
+    final surveyDetails = ref.watch(_surveyDetailsStreamProvider).value;
+    final errorMessage = ref.watch(_errorStreamProvider).value ?? "";
+    return ref.watch(formViewModelProvider).when(
+          init: () => _buildFormScreen(isLoading: true),
+          loading: () => _buildFormScreen(isLoading: true),
```
Comment (Shayokh144) [inline lib/ui/form/form_screen.dart:54]: @Tuubz It seems init and loading both having same value for `isLoading` in that case should we remove the `init` state?

File: lib/ui/form/form_screen.dart [hunk 2]
```diff
@@ -1,44 +1,126 @@
...
 
+  @override
+  FormScreenState createState() => FormScreenState();
+}
+
+class FormScreenState extends ConsumerState<FormScreen> {
+  final _pageController = PageController();
+
+  @override
+  void initState() {
```
Comment (Shayokh144) [inline lib/ui/form/form_screen.dart:43]: @Tuubz should we override `dispose()` method in this screen?

File: test/ui/form/form_view_model_test.dart [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import 'package:flutter_survey/ui/form/form_state.dart';
+import 'package:flutter_survey/ui/form/form_view_model.dart';
+import 'package:flutter_survey/usecases/base/base_use_case.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+
+import '../../mocks/generate_mocks.mocks.dart';
+
+void main() {
+  group('FormViewModelTest', () {
```
Comment (Shayokh144) [inline test/ui/form/form_view_model_test.dart:14]: Also no test for `init` state, I think we can remove init state and start with loading state.

**Review by Shayokh144 — COMMENTED** [review]
Added 3 comments, please check

File: lib/ui/form/form_screen.dart (hunk 2 above)
Comment (Tuubz) [reply lib/ui/form/form_screen.dart:43]: Sounds good, but what will we dispose tho? 💭

**Comment by Tuubz** [comment]
@Shayokh144 @luongvo Updated 🚀

**Comment by Tuubz** [comment]
@Shayokh144 I've updated the code again, I think it's best to keep it.

Why? For example in `form_view_model_test.dart`, after your suggestion, the test must be updated from:

```
emitsInOrder([
  const FormState.loading(),
  const FormState.loadSurveyDetailsSuccess(),
]),
```

to this:

```
emitsInOrder([
  const FormState.loadSurveyDetailsSuccess(),
]),
```

This is weird because we set the state for `FormState.loading()` inside `loadSurveyDetails(...)`

Not sure how to fix this, thoughts? 💭

**Comment by Shayokh144** [comment]
> @Shayokh144 I've updated the code again, I think it's best to keep it.
> 
> Why? For example in `form_view_model_test.dart`, after your suggestion, the test must be updated from:
> 
> ```
> emitsInOrder([
>   const FormState.loading(),
>   const FormState.loadSurveyDetailsSuccess(),
> ]),
> ```
> 
> to this:
> 
> ```
> emitsInOrder([
>   const FormState.loadSurveyDetailsSuccess(),
> ]),
> ```
> 
> This is weird because we set the state for `FormState.loading()` inside `loadSurveyDetails(...)`
> 
> Not sure how to fix this, thoughts? 💭

@Tuubz I think we have a misunderstanding here. I told you that we can remove the `init state` as `init and loading`is same according to code.

**Comment by Tuubz** [comment]
> > @Shayokh144 I've updated the code again, I think it's best to keep it.
> > Why? For example in `form_view_model_test.dart`, after your suggestion, the test must be updated from:
> > ```
> > emitsInOrder([
> >   const FormState.loading(),
> >   const FormState.loadSurveyDetailsSuccess(),
> > ]),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to this:
> > ```
> > emitsInOrder([
> >   const FormState.loadSurveyDetailsSuccess(),
> > ]),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is weird because we set the state for `FormState.loading()` inside `loadSurveyDetails(...)`
> > Not sure how to fix this, thoughts? 💭
> 
> @Tuubz I think we have a misunderstanding here. I told you that we can remove the `init state` as `init and loading`is same according to code.

Yes, correct. I tried removing the init state completely, but as a result we must update the test too 🙇 

Since I can't find a good solution for it, I'm proposing this [alternative](https://github.com/nimblehq/ic-flutter-taher-toby/pull/67/commits/4fbfe2530488ede91f85909c8b8cc3640cfd9b59), what do you think? 🙏

File: lib/ui/form/form_screen.dart [hunk 4]
```diff
@@ -1,44 +1,126 @@
...
-    // TODO: Integrate with VM #23
-    _getSurveyDetailsUseCase.call(
-      GetSurveyDetailsInput(
-        surveyId: surveyId,
-      ),
-    );
+    final surveyDetails = ref.watch(_surveyDetailsStreamProvider).value;
+    final errorMessage = ref.watch(_errorStreamProvider).value ?? "";
+    return ref.watch(formViewModelProvider).when(
+          init: () => Container(),
```
Comment (Shayokh144) [inline lib/ui/form/form_screen.dart:53]: Is it necessary have an empty container? we can remove the `init state` completely, our first initial state can be the `loading state`. What do you think?

**Review by Shayokh144 — COMMENTED** [review]
Similar logic to previous comment

**Review by Shayokh144 — APPROVED** [review]

File: lib/ui/form/form_screen.dart [hunk 5]
```diff
@@ -1,44 +1,126 @@
...
+                if (index == 0) {
+                  return FormSurveyDetailPage(
+                    surveyDetails: surveyDetails,
+                  );
+                } else {
+                  return const FormSurveyQuestionPage();
+                }
+              },
+            ),
+            _buildStartButton()
```
Comment (luongvo) [inline lib/ui/form/form_screen.dart:91]: Does this mean we will have to show this Start Button for all question pages? Should this button remain on the `FormSurveyDetailPage`?

Comment (luongvo) [reply lib/ui/form/form_screen.dart:91]: or are you planning to show/hide the corresponding button here by the `index`?

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: lib/ui/form/form_screen.dart (hunk 5 above)
Comment (Tuubz) [reply lib/ui/form/form_screen.dart:91]: > or are you planning to show/hide the corresponding button here by the index?

Yes, correct. I've integrated this in the [next PR](https://github.com/nimblehq/ic-flutter-taher-toby/pull/71/files#diff-74190efa1f96d56465cdfdd4ed8a7df6402195aab6a35890f6b57d70027df792R107-R110) 🚀 

What do you think of this approach? 🙏

Comment (luongvo) [reply lib/ui/form/form_screen.dart:91]: @Tuubz That's fine 👍  Putting the button on each page will produce the Button that will be moving with the page when scrolling the pager index with animation. Putting the button outside of and on the page will not produce that result. Just a small different thing here, it's up to you to decide which one is the best, even making changes in the following PRs.

**Review by luongvo — APPROVED** [review]

**Pull request merged**